optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -460,12 +460,20 @@ public void beforeProblemPropertyChanged(Object problemFactOrEntity) {
 
     @Override
     public void afterProblemPropertyChanged(Object problemFactOrEntity) {
-        variableListenerSupport.resetWorkingSolution(); // TODO do not nuke the variable listeners
+        if (isConstraintConfiguration(problemFactOrEntity)) {
+            setWorkingSolution(workingSolution); // Nuke everything and recalculate, constraint weights have changed.
+        } else {
+            variableListenerSupport.resetWorkingSolution(); // TODO do not nuke the variable listeners
+        }
     }
 
     @Override
     public void beforeProblemFactRemoved(Object problemFact) {
-        // Do nothing
+        if (isConstraintConfiguration(problemFact)) {
+            throw new IllegalStateException("Attempted to remove constraint configuration (" + problemFact +
+                    ") from solution (" + workingSolution + ").\n" +
+                    "Maybe use before/afterProblemPropertyChanged(...) instead.");
+        }
```

Comment:
In theory you could remove a constraint configuration and then add in another one.
however, if we nuke the working memory in between, that's going to cause issues.
Ideally, we should then delay the nuking which is very complex.

Long story short: +1 to do it like this for now, maybe add a todo comment